w-12273311-Reuse types first draft (#235)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-type-reuse-xml-schema.adoc

= Reusing an XML Schema
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

You can import XML schema files (`.xsd`) inside your DataWeave script as modules by using the `xmlschema!` module loader. This loader enables you to use types that are declared on your schema inside DataWeave directly.
DataWeave loads your XSD file and translates declarations inside your file into DataWeave type directives that can be accessed in the same way as types from any other DataWeave module. Use them to build new types, type check your variables, do pattern matching, or declare new functions that use types. There are no restrictions on how to use these types.

> The following example shows how to look for an XSD file in the path `org/weave/myfolder/User.xsd` being the first path the root of the classpath of your application:

Reviewer: Hmmm, not sure about that. Maybe we could change it to :
`The following example shows how to look for an XSD file in the path `org/weave/myfolder/User.xsd` inside your DataWeave script:`

----

The following example shows how to import a type:

----
import * from xmlschema!example::schema::Person
----

== Use Your Types in a DataWeave Script

This example shows how to use types to import XML schema files into your DataWeave script.

.XSD File (`User.xsd`):
[source,XML]
----
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified"
           xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="Root">
        <xs:complexType>
            <xs:choice>
                <xs:element name="employee" type="employee"/>
                <xs:element name="member" type="member"/>
            </xs:choice>
            <xs:attribute name="id" use="required"/>
        </xs:complexType>
    </xs:element>
    <xs:complexType name="employee">
        <xs:sequence>
            <xs:element name="name" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>

    <xs:complexType name="member">
        <xs:sequence>
            <xs:element name="id" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>
</xs:schema>
----

.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0

> import * from xmlschema!org::mule::weave::v2::module::xmlschema::engine::pattern_matching::User

Author: Where does User come from in this? xmlschema!org::mule::weave::v2::module::xmlschema::engine::pattern_matching::User

I don't see it in the schema, for example. I see Root as the element name.
Does this example point to a path in our source code? Should we use that path?
@martincousido

Reviewer: That path is the location of the schema file, it doesn't appear inside it.
It is a test that he have in our repo. I think we could change it to:
```suggestion
import * from xmlschema!org::weave::myfolder::User
```
That is the same as the first example at the beginning of this page

{
      Root @(id:"test") :{
        employee: {
            "name": "Mariano"
        }
      } match {
             case is Root -> true
             else -> false
        }
}
----

> == From the XML Model to DataWeave

Author: Can we make this title clearer? I make a suggestion in another comment below
@martincousido

Reviewer: Maybe "From the XML Schema Model to DataWeave types"

> As previously mentioned, the XML schema type system translates to DataWeave type directives. Therefore, as the two type models are considered different, some general rules and limitations apply.

Author: "..., some general rules and limitations apply."
We still need to work on this. As customers might ask/request those rules and limitations.
@martincousido

Reviewer: The rules and limitations are everything that follows until the end of that page

Author: What do you think of rephrasing the sentence to the following?   
"Therefore, as the two type models are considered different, the resulting DataWeave Type might not represent exactly the same XML Model."
@TomoDC

Reviewer: Looks good to me

Author: Reworded as mentioned as discussed with Christian in Slack.

> The following files illustrate how XSD translates to DataWeave code. These files are generated dynamically and are not visible, though they are useful for learning how to consume types from generated XSD files.

Author: "The following files..."
If we are referring to the "Built-in XML Schema Types", we might want to delete that title and add those types directly under this paragraph. 

If we do that, can we rephrase "From the XML Model to DataWeave" to "Built-in XML Schema Types"? What do you think?
@martincousido

Reviewer: I think that the "From the XML Model to DataWeave" covers everything that follows. So for example this title should have "XSD Structure mapping" also inside it (it should be "===" not a "==")

I wouldn't delete the title "Built-in XML Schema Types", I would change "=== XSD Structure mapping"
To make it a little more clearer we could change it to "From the XML Schema Model to DataWeave types"

Reviewer: Or maybe another options is to delete the "== XSD Structure mapping" title and just continue everything as it is (sub sections of "From the XML Model to DataWeave" So "Built-in XML Schema Types" is at the same level as "=== Root Elements Declaration", "=== Complex Types", and so on)

                <xs:element name="float" type="xs:float"/>
                <xs:element name="double" type="xs:double"/>

                <xs:element name="date" type="xs:date"/>
                <xs:element name="dateTime" type="xs:dateTime"/>
                <xs:element name="time" type="xs:time"/>

                <xs:element name="boolean" type="xs:boolean"/>

                <xs:element name="hexBinary" type="xs:hexBinary"/>
                <xs:element name="base64Binary" type="xs:base64Binary"/>
                <xs:element name="decimal" type="xs:decimal"/>

            </xs:all>
        </xs:complexType>
    </xs:element>
</xs:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

type RootElementDefinition = {| Root: { string: String, normalizedString: String, integer: Number, negativeInteger: Number, int: Number, unsignedInt: Number, unsignedLong: Number, unsignedShort: Number, float: Number, double: Number, date: LocalDateTime, dateTime: DateTime, time: Time, boolean: Boolean, hexBinary: Binary, base64Binary: Binary, decimal: Number } |}

type Root = RootElementDefinition
----

== XSD Structure Mapping

> 

Author: We should add a short intro here. 
@martincousido

Reviewer: I would delete that title so the following titles are inside "From the XML Model to DataWeave"

=== Root Elements Declaration

> Each root element declared inside your XSD file generates a type directive with the element name followed by the `ElementDeclaration` extension and a type directive named *Root*, considered a Union type, consisting of all your root elements.

Author: Making a suggestion here. What do you think?

Each root element (`<xs:element/>`) declared inside your XSD file generates the following:
* A DataWeave `type` directive with the element name followed by the `ElementDeclaration` extension.
* A `type` directive named `Root` as a `Union` type that consists of all the root elements in the XSD.
@martincousido

Reviewer: Yes! I agree with that suggestion. The only change would be to put <xs:element> instead of <xs:element/> (without the "/" slash)

= Reusing an XML Schema
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> You can import XML schema files (`.xsd`) inside your DataWeave script as modules by using the `xmlschema!` module loader. This loader enables you to use types that are declared on your schema inside DataWeave directly.

Reviewer 2: "in" or "on" your schema?

Author: in has more hits.

> DataWeave loads your XSD file and translates declarations inside your file into DataWeave type directives that can be accessed in the same way as types from any other DataWeave module. Use them to build new types, type check your variables, do pattern matching, or declare new functions that use types. There are no restrictions on how to use these types.

Reviewer 2: ```suggestion
DataWeave loads your XSD file and translates declarations in your file into DataWeave type directives that you can access in the same way as types from any other DataWeave module. Use the directives to build new types, type-check your variables, match patterns, or declare new functions that use types. DataWeave places no restrictions on how to use these types.
```

Author: Reworded.

The following example shows how to look for an XSD file in the path `org/weave/myfolder/User.xsd` inside your DataWeave script:

----
import * from xmlschema!org::weave::myfolder::User
----

== Import Syntax

> To import the XML schema types, as shown above, use the following syntax, where:

Reviewer 2: avoid above

Author: Reworded.

> * `typeToImport`: You can use `*` to import all types defined in the schema, or to import a single type from the schema, for example, `Root`. You can also import XML schema types with a different name, for example, `Root as Person`. This way  you can reference the type with that name in the script.

Reviewer 2: can you make this list a partial and reuse it in https://github.com/mulesoft/docs-dataweave/pull/235/files#diff-81a6acd824ff907819c4bee8ad9365b07c47760a3074d9230b13754b0131b4bc?

> * `typeToImport`: You can use `\*` to import all types defined in the schema, or to import a single type from the schema, for example, `Root`. You can also import XML schema types with a different name, for example, `Root as Person`. This way  you can reference the type with that name in the script.

Reviewer: ```suggestion
* `typeToImport`: You can use `*` to import all types defined in the schema, or to import a single type from the schema, for example, `Root`. You can also import XML schema types with a different name, for example, `Root as Person`. This way  you can reference the type with that name in the script.
```
The \ is rendered when viewing the file in GitHub. Not sure if this will happen also in the docs page.

Author: Modiefied.

[source,XML]
----
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified"
           xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="Root">
        <xs:complexType>
            <xs:choice>
                <xs:element name="employee" type="employee"/>
                <xs:element name="member" type="member"/>
            </xs:choice>
            <xs:attribute name="id" use="required"/>
        </xs:complexType>
    </xs:element>
    <xs:complexType name="employee">
        <xs:sequence>
            <xs:element name="name" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>

    <xs:complexType name="member">
        <xs:sequence>
            <xs:element name="id" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>
</xs:schema>
----

.DataWeave Script:
[source,dataweave,linenums]
----
%dw 2.0
import * from xmlschema!org::weave::myfolder::User

{
      Root @(id:"test") :{
        employee: {
            "name": "Mariano"
        }
      } match {
             case is Root -> true
             else -> false
        }
}
----

== From the XML Model to DataWeave Types

> As previously mentioned, the XML schema type system translates to DataWeave type directives. Therefore, as the two type models are considered different, the resulting DataWeave Type might not represent exactly the same XML Model.

Reviewer 2: Avoid "previously". If there's more info about it somewhere else, link to it.

Should Type be lc here?

> As previously mentioned, the XML schema type system translates to DataWeave type directives. Therefore, as the two type models are considered different, the resulting DataWeave Type might not represent exactly the same XML Model.

Reviewer 2: ```suggestion
Because the XML schema type system translates to DataWeave type directives and the two type models are different, the resulting DataWeave type might not represent exactly the same XML model.
```

Author: Reworded.

        <xs:complexType>
            <xs:all>
                <xs:element name="string" type="xs:string"/>
                <xs:element name="normalizedString" type="xs:normalizedString"/>

                <xs:element name="integer" type="xs:integer"/>
                <xs:element name="negativeInteger" type="xs:negativeInteger"/>
                <xs:element name="int" type="xs:int"/>
                <xs:element name="unsignedInt" type="xs:unsignedInt"/>
                <xs:element name="unsignedLong" type="xs:unsignedLong"/>
                <xs:element name="unsignedShort" type="xs:unsignedShort"/>
                <xs:element name="float" type="xs:float"/>
                <xs:element name="double" type="xs:double"/>

                <xs:element name="date" type="xs:date"/>
                <xs:element name="dateTime" type="xs:dateTime"/>
                <xs:element name="time" type="xs:time"/>

                <xs:element name="boolean" type="xs:boolean"/>

                <xs:element name="hexBinary" type="xs:hexBinary"/>
                <xs:element name="base64Binary" type="xs:base64Binary"/>
                <xs:element name="decimal" type="xs:decimal"/>

> 

Reviewer 2: extra space ok?

            </xs:all>
        </xs:complexType>
    </xs:element>
</xs:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

type RootElementDefinition = {| Root: { string: String, normalizedString: String, integer: Number, negativeInteger: Number, int: Number, unsignedInt: Number, unsignedLong: Number, unsignedShort: Number, float: Number, double: Number, date: LocalDateTime, dateTime: DateTime, time: Time, boolean: Boolean, hexBinary: Binary, base64Binary: Binary, decimal: Number } |}

type Root = RootElementDefinition
----

=== Root Elements Declaration

> Each root element (`<xs:element>`) declared inside your XSD file generates the following:

Reviewer 2: inside->in?

Author: Reworded.

> * A DataWeave `type` directive with the element name followed by the `ElementDefinition` extension.

Reviewer 2: no terminating punctuation if all bullets are fragments

Author: Deleting full stop.

> * A DataWeave `type` directive with the element name followed by the `ElementDeclaration` extension.

Reviewer: ```suggestion
* A DataWeave `type` directive with the element name followed by the `ElementDefinition` extension.
```
It should be ElementDefinition

Author: Modified.

.XSD:
[source,XML]
----
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">

    <xs:element name="Root">
        <xs:complexType>
            <xs:sequence>
                <xs:element type="xs:string" name="color" default="Red"/>
                <xs:element type="xs:integer" name="age" default="12"/>
            </xs:sequence>
            <xs:attribute type="xs:string" name="type" default="ALBA"/>
        </xs:complexType>
    </xs:element>

    <xs:element name="AnotherRoot">
        <xs:complexType>
            <xs:sequence>
                <xs:element type="xs:string" name="name" default="Red"/>
            </xs:sequence>
            <xs:attribute type="xs:string" name="type"/>
        </xs:complexType>
    </xs:element>
</xs:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

type AnotherRootElementDefinition = {| AnotherRoot @("type"?: String): {- name: String -} |}

type RootElementDefinition = {| Root @("type"?: String): {- color: String, age: Number -} |}

type Root = AnotherRootElementDefinition | RootElementDefinition
----

=== Complex Types

> Named complex types declared as top-level elements on the schema which are referenced inside your schema or are available for reuse on other structures generate their own type directive. These type directives are named with the `complextype` name followed by the `Definition` extension.

Reviewer 2: Long drink of water...

Named complex types (a bunch of modifiers) generate their own type directive.

What about:

Named complex types generate their own type directive when:
* blah
* dee
* dah

> Named complex types declared as top-level elements on the schema which are referenced inside your schema or are available for reuse on other structures generate their own type directive. These type directives are named with the `complextype` name followed by the `Definition` extension.

Reviewer 2: ```suggestion
Named complex types declared as top-level elements on the schema which are referenced inside your schema or are available for reuse on other structures generate their own type directive. These type directive names include `complextype` followed by the `Definition` extension.
```

Author: Modified.

<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified"
           xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="Root">
        <xs:complexType>
            <xs:choice>
                <xs:element name="employee" type="employee"/>
                <xs:element name="member" type="member"/>
            </xs:choice>
            <xs:attribute name="id" default="foo"/>
        </xs:complexType>
    </xs:element>
    <xs:complexType name="employee">
        <xs:sequence>
            <xs:element name="name" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>

    <xs:complexType name="member">
        <xs:sequence>
            <xs:element name="id" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>
</xs:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

type EmployeeDefinition = { employee: {- name: String -} }

type MemberDefinition = { member: {- id: String -} }

type RootElementDefinition = {| Root @(id?: String): {| employee: EmployeeDefinition.employee |} | {| member: MemberDefinition.member |} |}

type Root = RootElementDefinition
----

=== Imports/Includes

> The use of imports or includes inside an XSD schema file results in a DataWeave module which contains an `import` directive that uses the XML schema module loader to point to the schema file located at the specified `schemaLocation`. This location must be declared relative to the main schema. Notice how the following examples use xref:dataweave-selecting-types.adoc[]:

Reviewer 2: global - look at "inside" to see if "in" will work.

Reviewer 2: This is a long sentence - any way to split?

> The use of imports or includes inside an XSD schema file results in a DataWeave module which contains an `import` directive that uses the XML schema module loader to point to the schema file located at the specified `schemaLocation`.

.XSD:
[source,XML]
----
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema"
						xmlns:common="http://common.com/COMMON">

	<xsd:import schemaLocation="common.xsd" namespace="http://common.com/COMMON" />

	<xsd:element name="Root" type="common:RequestHeader"/>

</xsd:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

ns ns0 http://common.com/COMMON

import xmlschema!org::mule::weave::v2::module::xmlschema::moduleloader::imports::common

type RootElementDefinition = {| Root: common::RequestHeaderDefinition.ns0#RequestHeader |}

type Root = RootElementDefinition
----

=== Choice

> You can map this XML schema structure to a `UnionType` in DataWeave:

Reviewer 2: a UnionType what?

Author: Deleting indef art.

<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified"
           xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="Root">
        <xs:complexType>
            <xs:choice>
                <xs:element name="employee" type="employee"/>
                <xs:element name="member" type="member"/>
            </xs:choice>
        </xs:complexType>
    </xs:element>
    <xs:complexType name="employee">
        <xs:sequence>
            <xs:element name="name" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>

    <xs:complexType name="member">
        <xs:sequence>
            <xs:element name="id" type="xs:string"/>
        </xs:sequence>
    </xs:complexType>
</xs:schema>
----

.Generated DataWeave Module:
[source,dataweave,linenums]
----
%dw 2.0

type EmployeeDefinition = { employee: {- name: String -} }

type MemberDefinition = { member: {- id: String -} }

type RootElementDefinition = {| Root: {| employee: EmployeeDefinition.employee |} | {| member: MemberDefinition.member |} |}

type Root = RootElementDefinition
----

=== Namespaces

Namespaces in your XML schema file translate directly to DataWeave type directives.
 

> The following example shows how to set the specific `targetNamespace` where your elements exist. This namespace, which translates to DataWeave, is required if you want to select and navigate said types. 

Reviewer 2: ```suggestion
The following example shows how to set the specific `targetNamespace` <thingie> where your elements exist. This namespace, which translates to DataWeave, is required to select and navigate these types. 
```

Author: Modified.
